### ISSUE TITLE
Use Fuzzy Search?

### ISSUE THREAD

**Issue by JP-Ellis** [brunnerh/insert-unicode#13]
It would be good if the search was 'fuzzy', or at least less restrictive.

For example, I would expect that the search for `greek beta` would give all results which have both `greek` and `beta` in their description; however, the extension only returns matches which contain `greek beta` as an exact substring.

I realize that it's possible to use `greek*beta` which is a partial solution, it is still sensitive to the order of words (for example, `beta*greek` returns no result).

**Comment by brunnerh** [comment]
That would be quite useful, but the extension has to rely on the available mechanisms. In this case a "quick-pick" which currently [has no option for that](https://code.visualstudio.com/api/references/vscode-api#QuickPickOptions).

**Comment by JP-Ellis** [comment]
That's a bit of shame.

Would it be possible to make multiple calls to the quick-pick API and combine the results?  So for example, searching for `greek beta` would make two calls: `greek*beta` and `beta*greek` and combine the results?

This should work fine for 2 and 3 word searches (the latter requiring 6 API calls which might slow things down, but should also return very few results).  This is obviously not a scalable solution as the number of API calls required would grow factorially with the number of words, but I think it would be unlikely that more than 2 words is often necessary, let alone 3.

Alternatively, are there other VSCode extensions which could act as a drop-in replacement to the quick-pick call?

And as an interim 'fix', is it possible that it replaces all spaces with `*` automatically?

**Comment by brunnerh** [comment]
That is unfortunately not how the API works. The dropdown list is a unit and automatically filteres upon input with no way to change its logic apart from modifying where it searches (items have a label, optional description and details).

If there is no other way i might consider the automatic replacement as a configuration option.

**Comment by JP-Ellis** [comment]
Is it possible to use a similar API to the search use by `workbench.action.showAllSymbols`?  That search is fuzzy, and uses the same (at least visually) search box.

**Comment by brunnerh** [comment]
That appears to use some internals that are not exposed in the public API available to extensions.

There are already various issues open in the VS repository for functionality that would make it possible to implement a more lenient search. I will keep an eye on how that develops.

**Comment by fognome** [comment]
any update on this?

**Comment by brunnerh** [comment]
@fognome No, see [upstream issue](https://github.com/microsoft/vscode/issues/90521).

**Comment by lionel-rowe** [comment]
Are there alternative, lower-level UI components that could be used instead of `QuickPick`? Input noticeably lags on my machine, which could be mitigated by running the search logic in workers, or throttling input, or virtualizing the list elements (if that isn't done already), or some combination of the above. It seems like `QuickPick` isn't really intended for searching through tens of thousands of items as it doesn't have these customization options.

As well as performance and fuzziness, matching by alternate names would also be helpful, e.g. I expected to see `0x7d MULTIPLICATION SIGN` as an option when searching for `times`.

**Comment by brunnerh** [comment]
There are no alternatives that I know of. The quick pick is virtualized, otherwise it would be way worse (it used to not be virtualized). A proposed option for delaying the search is already part of the issue I opened.

The extension has an option to use an initial search followed by paging in the quick pick, you could try using that; the downside is lack of live feedback on the initial search and having to deal with pages.
```jsonc
"insert-unicode.disable-pre-filtering": false,
"insert-unicode.page-size": 500, // default, you can probably go higher here
```

For matching, there are alternatives for *some* characters based on a Unicode-provided list, some characters have quite a few but most have none. E.g. line feed (`0xA`):
> LINE FEED, NEW LINE, END OF LINE, LF, NL, EOL

I do not intend to extend this at the moment, though. The list of characters is huge and often this will be a one-time problem, eventually the character is found and the user now knows what to look for (and the character is at least locally added to the recently used list). Adding additional aliases may also cause characters to show up when people are looking for something else, so it can also potentially degrade the experience.

Would recommend using the [favorites](https://github.com/brunnerh/insert-unicode/?tab=readme-ov-file#insert-from-favorites) for characters you need frequently.

**Comment by AzzaAzza69** [comment]
I take it this is why "arrow down" shows 5 entries whilst "down arrow" shows 26?

**Comment by brunnerh** [comment]
The word order still appears to matter, though additional text can appear in-between; if I remember correctly that did not use to be the case and you needed to insert `*`.
